File: src/main/java/gargoyle/calendar/util/convert/Converters.java
package gargoyle.calendar.util.convert;

import gargoyle.calendar.util.Tuple;

import java.text.MessageFormat;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Optional;
import java.util.function.Function;


public class Converters {

    private final Map<Tuple<Class, Class>, Function> converters;

    protected Converters() {
        converters = new LinkedHashMap<>();
    }

    public Converters(Map<Tuple<Class, Class>, Function> converters) {
        this.converters = new LinkedHashMap<>(converters);
    }

    protected final <S, T> Converters declareConverter(Class<S> sourceType, Class<T> targetType, Function<S, T> converter) {
        if (canConvert(sourceType, targetType)) {
            throw new IllegalArgumentException(MessageFormat.format("converter {0}->{1} already exists",
                    sourceType, targetType));
        }
        converters.put(new Tuple<>(sourceType, targetType), converter);
        return this;
    }

    public final <S, T> boolean canConvert(Class<S> sourceType, Class<T> targetType) {
        return converters.containsKey(new Tuple<Class, Class>(sourceType, targetType));
    }

    public final <S, T, V extends S> T convert(Class<T> targetType, V value) {
        if (value == null) return null;
        else {
            Class<?> valueClass = value.getClass();
            return convert((Class<S>) valueClass, targetType, value);
        }
    }

    public final <S, T, V extends S> T convert(Class<S> sourceType, Class<T> targetType, V value) {
        return value == null ? null : (sourceType == targetType ? (T) value :
                (T) Optional.ofNullable(converters.get(new Tuple<Class, Class>(sourceType, targetType)))
                        .orElseThrow(() -> new UnsupportedOperationException(MessageFormat.format(
                                "converter {0}->{1} is not supported yet", sourceType, targetType)))
                        .apply(value));
    }
}

File: cal-core/src/main/kotlin/gargoyle/calendar/util/convert/Converters.kt
package gargoyle.calendar.util.convert

import kotlin.reflect.KClass

@Suppress("UNCHECKED_CAST")
open class Converters {
    private val converters: MutableMap<Pair<KClass<*>, KClass<*>>, (Any) -> Any>

    protected constructor() {
        converters = LinkedHashMap()
    }

    constructor(converters: Map<Pair<KClass<*>, KClass<*>>, (Any) -> Any>) {
        this.converters = LinkedHashMap(converters)
    }

    protected fun <S : Any, T : Any> declareConverter(
        sourceType: KClass<S>,
        targetType: KClass<T>,
        converter: (S) -> T
    ): Converters {
        require(!canConvert(sourceType, targetType)) { "converter ${sourceType}->${targetType} already exists" }
        converters[sourceType to targetType] = converter as (Any) -> Any
        return this
    }

    fun <S : Any, T : Any> canConvert(sourceType: KClass<S>, targetType: KClass<T>): Boolean {
        return (sourceType to targetType) in converters
    }

    @Suppress("UNNECESSARY_NOT_NULL_ASSERTION")
    fun <S : Any, T : Any, V : S?> convert(targetType: KClass<T>, value: V?): T? {
        return if (value == null) null
        else convert<S, T, V>(value!!::class as KClass<S>, targetType, value)
    }

    fun <S : Any, T : Any, V : S?> convert(sourceType: KClass<S>, targetType: KClass<T>, value: V?): T? =
        when {
            value == null -> null
            sourceType == targetType -> value as T
            else -> (requireNotNull(converters[sourceType to targetType])
            { "converter ${sourceType}->${targetType} is not supported yet" } as (S) -> T)(value)
        }
}
